File: frontend/lib/dto/observation_dto.dart
class ObservationDTO {
  final int? id;
  final int? ownerId;
  final int? botanicalInfoId;
  final int? hikeSessionId;
  final String? hikeSessionName;
  final String? scientificName;
  final String? preferredCommonName;
  final DateTime observedAt;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? displayName;
  final String? trailName;
  final String? habitat;
  final String? notes;
  final double? latitude;
  final double? longitude;
  final double? accuracyMeters;
  final double? elevationMeters;
  final String locationPrivacy;
  final String status;
  final double? identificationConfidence;
  final String? identificationProvider;
  final String? clientReference;
  final List<String> imageIds;

  const ObservationDTO({
    this.id,
    this.ownerId,
    this.botanicalInfoId,
    this.hikeSessionId,
    this.hikeSessionName,
    this.scientificName,
    this.preferredCommonName,
    required this.observedAt,
    this.createdAt,
    this.updatedAt,
    this.displayName,
    this.trailName,
    this.habitat,
    this.notes,
    this.latitude,
    this.longitude,
    this.accuracyMeters,
    this.elevationMeters,
    this.locationPrivacy = 'PRIVATE',
    this.status = 'UNIDENTIFIED',
    this.identificationConfidence,
    this.identificationProvider,
    this.clientReference,
    this.imageIds = const [],
  });

  factory ObservationDTO.fromJson(Map<String, dynamic> json) {
    return ObservationDTO(
      id: json['id'] as int?,
      ownerId: json['ownerId'] as int?,
      botanicalInfoId: json['botanicalInfoId'] as int?,
      hikeSessionId: json['hikeSessionId'] as int?,
      hikeSessionName: json['hikeSessionName'] as String?,
      scientificName: json['scientificName'] as String?,
      preferredCommonName: json['preferredCommonName'] as String?,
      observedAt: DateTime.parse(json['observedAt'] as String),
      createdAt: json['createdAt'] == null
          ? null
          : DateTime.parse(json['createdAt'] as String),
      updatedAt: json['updatedAt'] == null
          ? null
          : DateTime.parse(json['updatedAt'] as String),
      displayName: json['displayName'] as String?,
      trailName: json['trailName'] as String?,
      habitat: json['habitat'] as String?,
      notes: json['notes'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      accuracyMeters: (json['accuracyMeters'] as num?)?.toDouble(),
      elevationMeters: (json['elevationMeters'] as num?)?.toDouble(),
      locationPrivacy: json['locationPrivacy'] as String? ?? 'PRIVATE',
      status: json['status'] as String? ?? 'UNIDENTIFIED',
      identificationConfidence:
          (json['identificationConfidence'] as num?)?.toDouble(),
      identificationProvider: json['identificationProvider'] as String?,
      clientReference: json['clientReference'] as String?,
      imageIds: (json['imageIds'] as List<dynamic>? ?? [])
          .map((value) => value.toString())
          .toList(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      if (botanicalInfoId != null) 'botanicalInfoId': botanicalInfoId,
      if (hikeSessionId != null) 'hikeSessionId': hikeSessionId,
      if (clientReference != null) 'clientReference': clientReference,
      'observedAt': observedAt.toUtc().toIso8601String(),
      if (displayName?.trim().isNotEmpty == true)
        'displayName': displayName!.trim(),
      if (trailName?.trim().isNotEmpty == true) 'trailName': trailName!.trim(),
      if (habitat?.trim().isNotEmpty == true) 'habitat': habitat!.trim(),
      if (notes?.trim().isNotEmpty == true) 'notes': notes!.trim(),
      if (latitude != null) 'latitude': latitude,
      if (longitude != null) 'longitude': longitude,
      if (accuracyMeters != null) 'accuracyMeters': accuracyMeters,
      if (elevationMeters != null) 'elevationMeters': elevationMeters,
      'locationPrivacy': locationPrivacy,
      'status': status,
      if (identificationConfidence != null)
        'identificationConfidence': identificationConfidence,
      if (identificationProvider != null)
        'identificationProvider': identificationProvider,
    };
  }

  String get bestDisplayName {
    for (final String? candidate in [
      displayName,
      preferredCommonName,
      scientificName,
    ]) {
      if (candidate?.trim().isNotEmpty == true) return candidate!.trim();
    }
    return 'Unidentified trail find';
  }
}

File: backend/src/main/java/com/github/mdeluise/plantit/observation/ObservationDTO.java
package com.github.mdeluise.plantit.observation;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import io.swagger.v3.oas.annotations.media.Schema;

@Schema(name = "Observation", description = "A wild plant encountered at a particular time and place.")
public class ObservationDTO {
    @Schema(accessMode = Schema.AccessMode.READ_ONLY)
    private Long id;
    @Schema(accessMode = Schema.AccessMode.READ_ONLY)
    private Long ownerId;
    private Long botanicalInfoId;
    @Schema(accessMode = Schema.AccessMode.READ_ONLY)
    private String scientificName;
    @Schema(accessMode = Schema.AccessMode.READ_ONLY)
    private String preferredCommonName;
    private Instant observedAt;
    @Schema(accessMode = Schema.AccessMode.READ_ONLY)
    private Instant createdAt;
    @Schema(accessMode = Schema.AccessMode.READ_ONLY)
    private Instant updatedAt;
    private String displayName;
    private String trailName;
    private String habitat;
    private String notes;
    private Double latitude;
    private Double longitude;
    private Double accuracyMeters;
    private Double elevationMeters;
    private String locationPrivacy;
    private String status;
    private Double identificationConfidence;
    private String identificationProvider;
    @Schema(accessMode = Schema.AccessMode.READ_ONLY)
    private List<String> imageIds = new ArrayList<>();


    public Long getId() {
        return id;
    }


    public void setId(Long id) {
        this.id = id;
    }


    public Long getOwnerId() {
        return ownerId;
    }


    public void setOwnerId(Long ownerId) {
        this.ownerId = ownerId;
    }


    public Long getBotanicalInfoId() {
        return botanicalInfoId;
    }


    public void setBotanicalInfoId(Long botanicalInfoId) {
        this.botanicalInfoId = botanicalInfoId;
    }


    public String getScientificName() {
        return scientificName;
    }


    public void setScientificName(String scientificName) {
        this.scientificName = scientificName;
    }


    public String getPreferredCommonName() {
        return preferredCommonName;
    }


    public void setPreferredCommonName(String preferredCommonName) {
        this.preferredCommonName = preferredCommonName;
    }


    public Instant getObservedAt() {
        return observedAt;
    }


    public void setObservedAt(Instant observedAt) {
        this.observedAt = observedAt;
    }


    public Instant getCreatedAt() {
        return createdAt;
    }


    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }


    public Instant getUpdatedAt() {
        return updatedAt;
    }


    public void setUpdatedAt(Instant updatedAt) {
        this.updatedAt = updatedAt;
    }


    public String getDisplayName() {
        return displayName;
    }


    public void setDisplayName(String displayName) {
        this.displayName = displayName;
    }


    public String getTrailName() {
        return trailName;
    }


    public void setTrailName(String trailName) {
        this.trailName = trailName;
    }


    public String getHabitat() {
        return habitat;
    }


    public void setHabitat(String habitat) {
        this.habitat = habitat;
    }


    public String getNotes() {
        return notes;
    }


    public void setNotes(String notes) {
        this.notes = notes;
    }


    public Double getLatitude() {
        return latitude;
    }


    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }


    public Double getLongitude() {
        return longitude;
    }


    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }


    public Double getAccuracyMeters() {
        return accuracyMeters;
    }


    public void setAccuracyMeters(Double accuracyMeters) {
        this.accuracyMeters = accuracyMeters;
    }


    public Double getElevationMeters() {
        return elevationMeters;
    }


    public void setElevationMeters(Double elevationMeters) {
        this.elevationMeters = elevationMeters;
    }


    public String getLocationPrivacy() {
        return locationPrivacy;
    }


    public void setLocationPrivacy(String locationPrivacy) {
        this.locationPrivacy = locationPrivacy;
    }


    public String getStatus() {
        return status;
    }


    public void setStatus(String status) {
        this.status = status;
    }


    public Double getIdentificationConfidence() {
        return identificationConfidence;
    }


    public void setIdentificationConfidence(Double identificationConfidence) {
        this.identificationConfidence = identificationConfidence;
    }


    public String getIdentificationProvider() {
        return identificationProvider;
    }


    public void setIdentificationProvider(String identificationProvider) {
        this.identificationProvider = identificationProvider;
    }


    public List<String> getImageIds() {
        return imageIds;
    }


    public void setImageIds(List<String> imageIds) {
        this.imageIds = imageIds;
    }
}
